### PR TITLE
Add com.gitlab.ColinDuquesnoy.MellowPlayer

### PR DESCRIPTION
**Pull request by ColinDuquesnoy**
Add [MellowPlayer](https://gitlab.com/ColinDuquesnoy/MellowPlayer) to flathub 

*Note about permissions*

We use ``--socket=session-bus`` and ``--socket=system-bus`` in order to get mpris to work. Let me know if there is a more "secure" solution.

### CODE REVIEW

File: com.gitlab.ColinDuquesnoy.MellowPlayer.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+sdk: org.kde.Sdk
+command: MellowPlayer
+finish-args:
+  - --share=ipc
+  - --socket=x11
+  - --socket=wayland
+  - --share=network
+  - --socket=pulseaudio
+  - --device=dri
+  - --socket=session-bus
```
Comment (barthalion) [inline com.gitlab.ColinDuquesnoy.MellowPlayer.yml:15]: Something like this should work for mpris:

```
    "--talk-name=org.mpris.MediaPlayer2.Player",
    "--own-name=org.mpris.MediaPlayer2.MellowPlayer.*"
```

File: com.gitlab.ColinDuquesnoy.MellowPlayer.yml [hunk 2]
```diff
@@ -0,0 +1,37 @@
+app-id: com.gitlab.ColinDuquesnoy.MellowPlayer
+runtime: org.kde.Platform
+runtime-version:  "5.15"
+base: io.qt.qtwebengine.BaseApp
+base-version: "5.15"
+sdk: org.kde.Sdk
+command: MellowPlayer
+finish-args:
+  - --share=ipc
+  - --socket=x11
```
Comment (barthalion) [inline com.gitlab.ColinDuquesnoy.MellowPlayer.yml:10]:
```suggestion
  - --socket=fallback-x11
```

File: com.gitlab.ColinDuquesnoy.MellowPlayer.yml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  - --socket=wayland
+  - --share=network
+  - --socket=pulseaudio
+  - --device=dri
+  - --socket=session-bus
+  - --socket=system-bus
+  - --filesystem=/tmp
```
Comment (barthalion) [inline com.gitlab.ColinDuquesnoy.MellowPlayer.yml:17]: Does it need to share `/tmp` with the host?

**Comment by barthalion** [comment]
bot, build com.gitlab.ColinDuquesnoy.MellowPlayer

**Comment by barthalion** [comment]
Build failed due to missing metainfo file, if it's part of the code, it might be installed in the wrong place or under wrong name.

**Comment by bilelmoussaoui** [comment]
> Build failed due to missing metainfo file, if it's part of the code, it might be installed in the wrong place or under wrong name.

Seems so
https://gitlab.com/ColinDuquesnoy/MellowPlayer/-/blob/master/src/main/share/metainfo/mellowplayer.appdata.xml

The metainfo (or any exported file) should be prefixed by the app-id. So it should be `com.gitlab.ColinDuquesnoy.MellowPlayer.metainfo.xml` 
The metainfo seems to have some missing tags that are required for Flathub. See https://github.com/flathub/flathub/wiki/AppData-Guidelines for details.

File: com.gitlab.ColinDuquesnoy.MellowPlayer.yml (hunk 1 above)
Comment (ColinDuquesnoy) [reply com.gitlab.ColinDuquesnoy.MellowPlayer.yml:15]: Thank you I'll try that

File: com.gitlab.ColinDuquesnoy.MellowPlayer.yml (hunk 3 above)
Comment (ColinDuquesnoy) [reply com.gitlab.ColinDuquesnoy.MellowPlayer.yml:17]: I'll check that

**Comment by ColinDuquesnoy** [comment]
> > Build failed due to missing metainfo file, if it's part of the code, it might be installed in the wrong place or under wrong name.
> 
> Seems so
> https://gitlab.com/ColinDuquesnoy/MellowPlayer/-/blob/master/src/main/share/metainfo/mellowplayer.appdata.xml
> 
> The metainfo (or any exported file) should be prefixed by the app-id. So it should be `com.gitlab.ColinDuquesnoy.MellowPlayer.metainfo.xml`
> The metainfo seems to have some missing tags that are required for Flathub. See https://github.com/flathub/flathub/wiki/AppData-Guidelines for details.

I'll fix that upstream. Thank you

File: com.gitlab.ColinDuquesnoy.MellowPlayer.yml (hunk 1 above)
Comment (ColinDuquesnoy) [reply com.gitlab.ColinDuquesnoy.MellowPlayer.yml]: That work fine.

File: com.gitlab.ColinDuquesnoy.MellowPlayer.yml (hunk 3 above)
Comment (ColinDuquesnoy) [reply com.gitlab.ColinDuquesnoy.MellowPlayer.yml]: I've removed --filesystem=/tmp

**Comment by ColinDuquesnoy** [comment]
bot, build com.gitlab.ColinDuquesnoy.MellowPlayer

**Comment by ColinDuquesnoy** [comment]
bot, build com.gitlab.ColinDuquesnoy.MellowPlayer

**Comment by ColinDuquesnoy** [comment]
bot, build com.gitlab.ColinDuquesnoy.MellowPlayer

**Comment by barthalion** [comment]
It started only once for me, I had to delete `/var/home/bpiotrowski/.var/app/com.gitlab.ColinDuquesnoy.MellowPlayer/data/MellowPlayer/MellowPlayer3/single-instance.lock` to make it work again.

Defaulting to hiding in tray may also be misleading as GNOME by default has no tray icons at all.

**Comment by ColinDuquesnoy** [comment]
> It started only once for me, I had to delete /var/home/bpiotrowski/.var/app/com.gitlab.ColinDuquesnoy.MellowPlayer/data/MellowPlayer/MellowPlayer3/single-instance.lock to make it work again.

How did you quit the application the first time? That shouldn't happen if you exit the application by using Menu -> Quit (or Ctrl+Q)...

> Defaulting to hiding in tray may also be misleading as GNOME by default has no tray icons at all.

There is a setting to disable that behaviour (F2->General -> Close application to system tray) for those who prefer the gnome behaviour. It's a sane default on all other desktops.

Does those two "problems" prevent from publishing the app? To me these looks like bugs that may/should be reported upstream and that are not related to flatpak.

**Comment by barthalion** [comment]
> How did you quit the application the first time? That shouldn't happen if you exit the application by using Menu -> Quit (or Ctrl+Q)...

I closed the window via equivalent of Alt+F4.

> There is a setting to disable that behaviour (F2->General -> Close application to system tray) for those who prefer the gnome behaviour. It's a sane default on all other desktops.

Being hidden in tray – maybe, being locked out of the application completely, not really. I assume all other applications like Zoom use D-Bus to query if application is already running. (Bilal, please correct me if I'm wrong!)

> Does those two "problems" prevent from publishing the app?

Not a blockers certainly but it might be confusing to regular users.

**Comment by ColinDuquesnoy** [comment]
> being locked out of the application completely, not really

I agree.

MellowPlayer is a single instance application (implemented with QLocalServer & QLocalSocket instead of dbus so that it also works on Windows). Starting a second instance of the application should bring back the first instance (even if closed with Alt+F4). 

I've checked this mechanism is still working: it works fine with native binary (except if you kill the process or use Ctrl+C but that is another issue) but it fails when using the flatpak. So in the end I think this is a bug that should be fixed before publishing.

**Comment by ColinDuquesnoy** [comment]
@barthalion The bug occurs because I removed ``--filesystem=/tmp`` from the flatpak manifest.

I forgot QLocalServer & QLocalSocket use ``/tmp/APP_Name`` for the local socket file. Do you think it is problematic to add ``--filesystem=/tmp`` back?

**Comment by barthalion** [comment]
Can you try with `--env=TMPDIR=/var/tmp` instead?

**Comment by gombosg** [comment]
#off
MellowPlayer! Lovely! I use this daily for Deezer. :slightly_smiling_face:

**Comment by ColinDuquesnoy** [comment]
> Can you try with `--env=TMPDIR=/var/tmp` instead?

I tried to add it to the flatpak yaml manifest, doesn't work. Any idea?

**Comment by barthalion** [comment]
I'll try to poke around tommorow(ish) and see. You might want to take a look at https://github.com/flathub/com.nextcloud.desktopclient.nextcloud (source code specifically, not the manifest) as I imagine it faces similar problems with locking etc..

**Comment by ColinDuquesnoy** [comment]
I did a bit more testing, ``--env=TMPDIR=/var/tmp`` actually works on recent distributions (ubuntu) but not on CentOS 7. Strangely, on ubuntu, the file is not created (ls /var/tmp/MellowPlayer3 fails) but communication over the local socket seems to work.

I've also found a way to create the socket file in the home directory (e.g ``.var/app/com.gitlab.ColinDuquesnoy.MellowPlayer/data/MellowPlayer/MellowPlayer3/single-instance.lock``), there the socket file is correctly created and communication also works fine. I will try it on CentOS 7 tomorrow. If that works I'll use that solution.

Another problem I've encountered is that Qt is unable to detect a stale lock file (e.g. after a crash) because the application PID is still the same (I've learned pid are namespaced in flatpak so this "normal" behavior, although a bit surprising) so I may need to find another way to detect that the application is already running (maybe using dbus)

**Comment by barthalion** [comment]
> I did a bit more testing, --env=TMPDIR=/var/tmp actually works on recent distributions (ubuntu) but not on CentOS 7. Strangely, on ubuntu, the file is not created (ls /var/tmp/MellowPlayer3 fails) but communication over the local socket seems to work.

It was fixed in flatpak 1.6.2 IIRC. If bubblewrap is SUID, glibc drops some environment variables such as TMPDIR; the work around would be to have a wrapper script that would set TMPDIR and `exec` the main binary instead.

> Another problem I've encountered is that Qt is unable to detect a stale lock file (e.g. after a crash) because the application PID is still the same (I've learned pid are namespaced in flatpak so this "normal" behavior, although a bit surprising) so I may need to find another way to detect that the application is already running (maybe using dbus)

Yeah, that's correct. If you are using Qt to set up tray icon, you also want to add

```
"--talk-name=org.kde.StatusNotifierWatcher",
"--own-name=org.kde.*"
```
…to finish-args.

**Comment by ColinDuquesnoy** [comment]
bot, build com.gitlab.ColinDuquesnoy.MellowPlayer

**Comment by ColinDuquesnoy** [comment]
@barthalion 

I've fixed the single instance applications issues: I fixed QLocalSocket/QLocalServer socket file path and used a native lock file instead of QLockFile (looking at the source code of nextcloud desktop was helpful, thanks for that :+1:)

**Comment by ColinDuquesnoy** [comment]
I also added ``--talk-name=org.freedesktop.Notifications`` for libnotify

**Pull request closed without merging**